### PR TITLE
Map iteration support for ForEach processor

### LINKED ISSUE #55215: [Feature Request][Foreach Ingest Processor] Add support for iterating over Hashmaps
**Describe the feature**:

There has been an increase in the need to do a single action over a hashmap of JSON object, which is currently being handled by painless. Here is a real scenario example, in which incoming data coming from a device that cannot be modified on the source side has certain fields mapped as strings instead of integers, preventing the possibility of aggregation and visualisation.

```
"docs": [
    {
      "_index": "index",
      "_id": "id",
      "_source": {
        "system": {
          "hostname": "somemachine",
          "diskStorage": {
            "/": {
              "1024-blocks": "234",
              "Capacity": "5%",
              "name": "/"
            },
            "/dev": {
              "1024-blocks": "654616",
              "Capacity": "23%",
              "name": "/dev"
            },
            "/dev/shm": {
              "1024-blocks": "5430",
              "Capacity": "543%",
              "name": "/dev/shm"
            }
          }
        }
      }
    }
  ]
```

If I would want to iterate over diskStorage here, and remove the % + changing the type from String to Int, trying to put this into a foreach processor like so:

```
"pipeline": {
    "description": "_description",
    "processors": [
      {
        "foreach": {
          "field": "system.diskStorage",
          "processor": {
            "gsub": {
              "field": "_ingest._value.Capacity",
              "pattern": "%",
              "replacement": ""
            }
          }
        }
      }
    ]
  },
```

Will return the below error:
`"type" : "illegal_argument_exception",
  "reason" : "field [system.diskStorage] of type [java.util.HashMap] cannot be cast to [java.util.List]"`

This is of course that this is not an array of objects, but it would be great to be able to do a foreach operation on hashmaps like these as well.

Currently as part of a script processor, you can do something like this as an example:

```
processors": [
      {
        "script": {
          "lang": "painless",
          "source": """
          def entryset = ctx.system?.diskStorage?.entrySet();
          if (entryset != null) {
            for (def x : entryset) {
              x.getValue().Capacity = Integer.parseInt(x.getValue().Capacity.replace('%', ''));
            }
          }
          """
        }
      }
```

### PR DESCRIPTION
**Pull request by danhermann**
Adds support to the ForEach processor for iterating over map fields in addition to array fields. When iterating over maps, an additional ingest metadata field `_ingest._key` is populated with the value of the map entry's key. Both the map entry's key and value can be changed. Map entries can also be removed completely by setting the key to a null or empty string value.

Resolves https://github.com/elastic/elasticsearch/issues/55215

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

File: modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/ForEachProcessorTests.java [hunk 1]
```diff
@@ -311,6 +312,56 @@ public void testRemovingFromTheSameField() {
...
+        });
+        ForEachProcessor processor = new ForEachProcessor("_tag", null, "field", testProcessor, true);
+        processor.execute(ingestDocument, (result, e) -> {});
+        assertThat(testProcessor.getInvokedCounter(), equalTo(3));
+        assertThat(encounteredKeys.toArray(), arrayContainingInAnyOrder("foo", "bar", "baz"));
+        assertThat(encounteredValues.toArray(), arrayContainingInAnyOrder(1, 2, 3));
+        assertThat(ingestDocument.getFieldValue("field", Map.class).entrySet().toArray(),
+            arrayContainingInAnyOrder(Map.entry("foo", 1), Map.entry("baz", 3)));
+    }
+
```
Comment (martijnvg) [inline modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/ForEachProcessorTests.java:364]: Maybe also add a nested for each tests for maps like the existing `testNestedForEach()` test?

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ForEachProcessor.java [hunk 2]
```diff
@@ -67,20 +67,61 @@ boolean isIgnoreMissing() {
...
+        Object o = ingestDocument.getFieldValue(field, Object.class, ignoreMissing);
+        if (o == null) {
             if (ignoreMissing) {
                 handler.accept(ingestDocument, null);
             } else {
                 handler.accept(null, new IllegalArgumentException("field [" + field + "] is null, cannot loop over its elements."));
             }
+        } else if (o instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<String, ?> map = (Map<String, ?>) o;
```
Comment (martijnvg) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ForEachProcessor.java:79]: I think we can cast to `Map<?, ?>` instead? Then the SuppressWarnings can be removed. In order to get the key a cast to String would required, but that shouldn't generate warnings.

**Review by martijnvg — COMMENTED** [review]
Looks good, left 2 comments.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@martijnvg, it took a long time 😞, but I fixed iteration over nested maps (two-line change [here](https://github.com/elastic/elasticsearch/pull/64062/commits/207982e8b374519ff9691c06a3e71f97f8f62a55#diff-d8bf82d8f0d8bfd971c61a0fc6a14671e1466e78951cbfc103b5de3fa882857c)) and added two test cases to verify that.

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by danhermann** [comment]
Thanks, @martijnvg!

**Pull request merged**